docs/caching_and_memoization.rst

```diff
@@ -0,0 +1,316 @@
+parameters.
+
+This document describes the use of these two utilities.
+
+
+Memoized
+--------
+
+Memoized is an in-memory cache. At its simplest, it's a replacement for the two
+common patterns used in this example class:
+
+.. code-block:: python
+
+   class MyClass(object):
+
+       def __init__(self):
+           self._all_objects = None
+           self._objects_by_key = {}
+
+       @property
+       def all_objects(self):
+           if self._all_objects is None:
+               result = do_a_bunch_of_stuff()
+               self._all_objects = result
+           return self._all_objects
+
+       def get_object_by_key(self, key):
+           if key not in self._objects_by_key:
+               result = do_a_bunch_of_stuff(key)
+               self._objects_by_key[result]
+           return self._objects_by_key[key]
+
+With the memoized decorator, this becomes:
+
+.. code-block:: python
+
+   from memoized import memoized
+
+   class MyClass(object):
+
+       @property
+       @memoized
+       def all_objects(self):
+           return do_a_bunch_of_stuff()
+
+       @memoized
+       def get_object_by_key(self, key):
+           return do_a_bunch_of_stuff(key)
+
+When decorating a class method, ``@memoized`` stores the results of calls to
+those methods on the class instance. It stores a result for every unique set of
+arguments passed to the decorated function. This persists as long as the class
+does (or until you manually invalidate), and will be garbage collected along
+with the instance.
+
+You can decorate any callable with ``@memoized`` and the cache will persist for
+the life of the callable. That is, if it isn't an instance method, the cache
+will probably be stored in memory for the life of the process. This should be
+used sparingly, as it can lead to memory leaks. However, this can be useful for
+lazily initializing singleton objects. Rather than computing at module load
+time:
+
+.. code-block:: python
+
+    def get_classes_by_doc_type():
+        # Look up all subclasses of Document
+        return result
+
+    classes_by_doc_type = get_classes_by_doc_type()
+    
+You can memoize it, and only compute if and when it's needed. Subsequent calls
+will hit the cache.
+
+.. code-block:: python
+
+    @memoized
+    def get_classes_by_doc_type():
+        # Look up all subclasses of Document
+        return result
+
+Quickcache
+----------
+
+``@quickcache`` behaves much more like a normal cache. It stores results in a
+caching backend (Redis, in CCHQ) for a specified timeout (5 minutes, by
+default). This also means they can be shared across worker machines. Quickcache
+also caches objects in local memory (10 seconds, by default). This is faster to
+access than Redis, but its not shared across machines.
+
+Quickcache requires you to specify which arguments to "vary on", that is, which
+arguments uniquely identify a cache
+
+For examples of how it's used, check out `the repo <repo>_`_. For background,
```

Comment:
I think there's a typo in the link syntax here

Reply:
Thanks, fixed